mlflow/models/evaluation/default_evaluator.py

```diff
@@ -614,21 +625,100 @@ def plot_pr_curve():
 
         self._log_pandas_df_artifact(per_class_metrics_collection_df, "per_class_metrics")
 
-    def _evaluate_custom_metrics(self):
+    def _log_custom_metric_artifact(self, artifact_name, raw_artifact, custom_metric_tuple):
+        if isinstance(raw_artifact, EvaluationArtifact):
+            return raw_artifact
+
+        exception_header = (
+            f"Custom metric function '{custom_metric_tuple.name}' at index "
+            "{custom_metric_tuple.index} in the `custom_metrics` parameter"
+        )
+
+        inferred_from_path, inferred_type, inferred_ext = _infer_artifact_type_and_ext(
+            artifact_name, raw_artifact, custom_metric_tuple
+        )
+        artifact_file_name = _gen_log_key(artifact_name, self.dataset_name) + inferred_ext
+        artifact_file_local_path = self.temp_dir.path(artifact_file_name)
+
+        # ParquetEvaluationArtifact isn't explicitly stated here because such artifacts can only
+        # be supplied through file. Which is handled by the first if clause. This is because
+        # DataFrame objects default to be stored as CsvEvaluationArtifact.
+        if inferred_from_path:
+            shutil.copyfile(raw_artifact, artifact_file_local_path)
+        elif inferred_type is JsonEvaluationArtifact:
+            with open(artifact_file_local_path, "w") as f:
+                if isinstance(raw_artifact, str):
+                    f.write(raw_artifact)
+                else:
+                    json.dump(raw_artifact, f)
+        elif inferred_type is CsvEvaluationArtifact:
+            raw_artifact.to_csv(artifact_file_local_path, index=False)
+        elif inferred_type is NumpyEvaluationArtifact:
+            np.save(artifact_file_local_path, raw_artifact, allow_pickle=False)
+        elif inferred_type is ImageEvaluationArtifact:
+            raw_artifact.savefig(artifact_file_local_path)
+        else:
+            # storing as pickle
+            try:
+                # Using protocol 4 because it is the highest supported protocol as of the
+                # introduction of PickleEvaluationArtifact. For more regarding pickle protocols,
+                # refer to: https://docs.python.org/3/library/pickle.html#data-stream-format
+                with open(artifact_file_local_path, "wb") as f:
+                    pickle.dump(raw_artifact, f, protocol=4)
```

Comment:
Can we reuse the `_save` routines that we already define for each of these artifact classes, rather than re-implementing them here?

Reply:
It would be tricky to do that. Because from my understanding, in order to construct an `EvaluationArtifact`-like object, you first need to save it to a local file, then call `mlflow.log_artifact`. In this case, we're saving user provided objects to local files. Is there ways around this?

Reply:
We should introduce a way for developers to construct before saving.

Reply:
I think this is a good idea since that's a better user experience when creating and logging custom `EvaluationArtifact` classes. Although I think this is pretty significant and could be out of scope of this PR. What do you all think about removing the ability for users to log ``EvaluationArtifact`` objects in this PR, and once I added the ability to construct `EvaluationArtifact` before saving I'll add back this functionality? @dbczumar @jinzhang21 @BenWilson2 @apurva-koti

Reply:
Agreed. I like the idea to provide general support on `_save`.

Reply:
Will remove the ability to log `EvaluationArtifact` for now. Once revisions to have general support on `_save` is done this ability will be added back.